### ISSUE TITLE
Support of AWD Bedrock Claude Sonnet v2 (cross region)

### ISSUE THREAD

**Issue by pstanisl** [BoltAI/BoltAI#220]
**Is your feature request related to a problem? Please describe.**

The new AWS Bedrock does not support cross regional Claude Sonnet v2. It would be nice to support it.

**Describe the solution you'd like**
Add support for Claude Sonnet v2 with model:

- `us.anthropic.claude-3-5-sonnet-20241022-v2:0`

**Additional context**
We have limited access only to cross regional model.

**Comment by longseespace** [comment]
Hey @pstanisl. I use [ListFoundationModels API](https://docs.aws.amazon.com/bedrock/latest/APIReference/API_ListFoundationModels.html) to fetch the list of models you can access directly from AWS SDK.

Looks like in your case, we will need to add support for custom [inference profile](https://docs.aws.amazon.com/bedrock/latest/userguide/inference-profiles.html). I can try to set this up but can you share more about your configuration. How did you setup Bedrock in your AWS Console?

**Comment by pstanisl** [comment]
Hey @longseespace, thank you for your response. I'm not expert in AWS (yet). I just opened AWS Bedrock in `us-east-1` where Claude Sonnet v2 is available as `cross-region inference` and asked for access.

![CleanShot 2024-12-07 at 08 20 56](https://github.com/user-attachments/assets/58241d88-4182-428a-aa38-fd3a38fd2a20)

Some info about it is [here](https://docs.aws.amazon.com/bedrock/latest/userguide/cross-region-inference.html).

As far a I know these cross region inference models are [quite new](https://aws.amazon.com/blogs/machine-learning/getting-started-with-cross-region-inference-in-amazon-bedrock/).

Here is the list of cross regional inference models in [docs](https://docs.aws.amazon.com/bedrock/latest/userguide/inference-profiles-support.html).

If you want to test anything or I should set anything, please let me know.

**Comment by pstanisl** [comment]
I was able to get the model `modelArn` when I called `list_inference_profiles`

```python
client = boto3.client("bedrock", region_name="us-east-1")
client.list_inference_profiles()
```

Response is this:

```python
{
    "ResponseMetadata": {
        "RequestId": "b9dda8a5-1b2a-4d1a-9e7a-10268ce1d8fd",
        "HTTPStatusCode": 200,
        "HTTPHeaders": {
            "date": "Sat, 07 Dec 2024 07:37:16 GMT",
            "content-type": "application/json",
            "content-length": "10143",
            "connection": "keep-alive",
            "x-amzn-requestid": "b9dda8a5-1b2a-4d1a-9e7a-10268ce1d8fd",
        },
        "RetryAttempts": 0,
    },
    "inferenceProfileSummaries": [
        {
            "inferenceProfileName": "US Anthropic Claude 3 Sonnet",
            "description": "Routes requests to Anthropic Claude 3 Sonnet in us-east-1 and us-west-2.",
            "createdAt": datetime.datetime(2024, 8, 26, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 8, 26, 0, 0, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.anthropic.claude-3-sonnet-20240229-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/anthropic.claude-3-sonnet-20240229-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/anthropic.claude-3-sonnet-20240229-v1:0"
                },
            ],
            "inferenceProfileId": "us.anthropic.claude-3-sonnet-20240229-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Anthropic Claude 3 Opus",
            "description": "Routes requests to Anthropic Cluade 3 Opus in us-east-1 and us-west-2.",
            "createdAt": datetime.datetime(2024, 8, 26, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 8, 26, 0, 0, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.anthropic.claude-3-opus-20240229-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/anthropic.claude-3-opus-20240229-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/anthropic.claude-3-opus-20240229-v1:0"
                },
            ],
            "inferenceProfileId": "us.anthropic.claude-3-opus-20240229-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Anthropic Claude 3 Haiku",
            "description": "Routes requests to Anthropic Claude 3 Haiku in us-east-1 and us-west-2.",
            "createdAt": datetime.datetime(2024, 8, 26, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 8, 26, 0, 0, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.anthropic.claude-3-haiku-20240307-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/anthropic.claude-3-haiku-20240307-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/anthropic.claude-3-haiku-20240307-v1:0"
                },
            ],
            "inferenceProfileId": "us.anthropic.claude-3-haiku-20240307-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Meta Llama 3.2 11B Instruct",
            "description": "Routes requests to Meta Llama 3.2 11B Instruct in us-east-1 and us-west-2.",
            "createdAt": datetime.datetime(2024, 9, 25, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 9, 25, 0, 0, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.meta.llama3-2-11b-instruct-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/meta.llama3-2-11b-instruct-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/meta.llama3-2-11b-instruct-v1:0"
                },
            ],
            "inferenceProfileId": "us.meta.llama3-2-11b-instruct-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Meta Llama 3.2 3B Instruct",
            "description": "Routes requests to Meta Llama 3.2 3B Instruct in us-east-1 and us-west-2.",
            "createdAt": datetime.datetime(2024, 9, 25, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 9, 25, 0, 0, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.meta.llama3-2-3b-instruct-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/meta.llama3-2-3b-instruct-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/meta.llama3-2-3b-instruct-v1:0"
                },
            ],
            "inferenceProfileId": "us.meta.llama3-2-3b-instruct-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Meta Llama 3.2 90B Instruct",
            "description": "Routes requests to Meta Llama 3.2 90B Instruct in us-east-1 and us-west-2.",
            "createdAt": datetime.datetime(2024, 9, 25, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 9, 25, 0, 0, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.meta.llama3-2-90b-instruct-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/meta.llama3-2-90b-instruct-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/meta.llama3-2-90b-instruct-v1:0"
                },
            ],
            "inferenceProfileId": "us.meta.llama3-2-90b-instruct-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Meta Llama 3.2 1B Instruct",
            "description": "Routes requests to Meta Llama 3.2 1B Instruct in us-east-1 and us-west-2.",
            "createdAt": datetime.datetime(2024, 9, 25, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 9, 25, 0, 0, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.meta.llama3-2-1b-instruct-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/meta.llama3-2-1b-instruct-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/meta.llama3-2-1b-instruct-v1:0"
                },
            ],
            "inferenceProfileId": "us.meta.llama3-2-1b-instruct-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Anthropic Claude 3.5 Sonnet",
            "description": "Routes requests to Anthropic Claude 3.5 Sonnet in us-east-1 and us-west-2.",
            "createdAt": datetime.datetime(2024, 8, 26, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 9, 26, 14, 9, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.anthropic.claude-3-5-sonnet-20240620-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/anthropic.claude-3-5-sonnet-20240620-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/anthropic.claude-3-5-sonnet-20240620-v1:0"
                },
            ],
            "inferenceProfileId": "us.anthropic.claude-3-5-sonnet-20240620-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Anthropic Claude 3.5 Sonnet v2",
            "description": "Routes requests to Anthropic Claude 3.5 Sonnet v2 in us-east-1, us-east-2 and us-west-2.",
            "createdAt": datetime.datetime(2024, 10, 22, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(
                2024, 10, 28, 21, 29, 55, 709743, tzinfo=tzutc()
            ),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.anthropic.claude-3-5-sonnet-20241022-v2:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/anthropic.claude-3-5-sonnet-20241022-v2:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-east-2::foundation-model/anthropic.claude-3-5-sonnet-20241022-v2:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/anthropic.claude-3-5-sonnet-20241022-v2:0"
                },
            ],
            "inferenceProfileId": "us.anthropic.claude-3-5-sonnet-20241022-v2:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Anthropic Claude 3.5 Haiku",
            "description": "Routes requests to US Anthropic Claude 3.5 Haiku in us-east-1, us-east-2 and us-west-2.",
            "createdAt": datetime.datetime(2024, 11, 4, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 11, 4, 0, 0, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.anthropic.claude-3-5-haiku-20241022-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/anthropic.claude-3-5-haiku-20241022-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-east-2::foundation-model/anthropic.claude-3-5-haiku-20241022-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/anthropic.claude-3-5-haiku-20241022-v1:0"
                },
            ],
            "inferenceProfileId": "us.anthropic.claude-3-5-haiku-20241022-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Meta Llama 3.1 8B Instruct",
            "description": "Routes requests to Meta Llama 3.1 8B Instruct in us-east-1, us-east-2 and us-west-2.",
            "createdAt": datetime.datetime(2024, 11, 1, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(
                2024, 11, 4, 18, 4, 3, 991725, tzinfo=tzutc()
            ),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.meta.llama3-1-8b-instruct-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/meta.llama3-1-8b-instruct-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-east-2::foundation-model/meta.llama3-1-8b-instruct-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/meta.llama3-1-8b-instruct-v1:0"
                },
            ],
            "inferenceProfileId": "us.meta.llama3-1-8b-instruct-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Meta Llama 3.1 70B Instruct",
            "description": "Routes requests to Meta Llama 3.1 70B Instruct in us-east-1, us-east-2 and us-west-2.",
            "createdAt": datetime.datetime(2024, 11, 1, 0, 0, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(
                2024, 11, 4, 18, 4, 24, 303028, tzinfo=tzutc()
            ),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.meta.llama3-1-70b-instruct-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/meta.llama3-1-70b-instruct-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-east-2::foundation-model/meta.llama3-1-70b-instruct-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/meta.llama3-1-70b-instruct-v1:0"
                },
            ],
            "inferenceProfileId": "us.meta.llama3-1-70b-instruct-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Nova Lite",
            "description": "Routes requests to Nova Lite in us-east-1, us-west-2 and us-east-2.",
            "createdAt": datetime.datetime(2024, 11, 29, 13, 23, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 11, 29, 13, 23, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.amazon.nova-lite-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/amazon.nova-lite-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/amazon.nova-lite-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-east-2::foundation-model/amazon.nova-lite-v1:0"
                },
            ],
            "inferenceProfileId": "us.amazon.nova-lite-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Nova Pro",
            "description": "Routes requests to Nova Pro in us-east-1, us-west-2 and us-east-2.",
            "createdAt": datetime.datetime(2024, 11, 29, 13, 23, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 11, 29, 13, 23, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.amazon.nova-pro-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/amazon.nova-pro-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/amazon.nova-pro-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-east-2::foundation-model/amazon.nova-pro-v1:0"
                },
            ],
            "inferenceProfileId": "us.amazon.nova-pro-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
        {
            "inferenceProfileName": "US Nova Micro",
            "description": "Routes requests to Nova Micro in us-east-1, us-west-2 and us-east-2.",
            "createdAt": datetime.datetime(2024, 11, 29, 13, 23, tzinfo=tzutc()),
            "updatedAt": datetime.datetime(2024, 11, 29, 13, 23, tzinfo=tzutc()),
            "inferenceProfileArn": "arn:aws:bedrock:us-east-1:339713155118:inference-profile/us.amazon.nova-micro-v1:0",
            "models": [
                {
                    "modelArn": "arn:aws:bedrock:us-east-1::foundation-model/amazon.nova-micro-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-west-2::foundation-model/amazon.nova-micro-v1:0"
                },
                {
                    "modelArn": "arn:aws:bedrock:us-east-2::foundation-model/amazon.nova-micro-v1:0"
                },
            ],
            "inferenceProfileId": "us.amazon.nova-micro-v1:0",
            "status": "ACTIVE",
            "type": "SYSTEM_DEFINED",
        },
    ],
}
```

I don't know if this is correct way. I will try to find out.

**Comment by pstanisl** [comment]
This [notebook](https://github.com/aws-samples/amazon-bedrock-workshop/blob/main/06_OpenSource_examples/cross-region-inference/Getting_started_with_Cross-region_Inference.ipynb) shows example how to use cross region models.

![CleanShot 2024-12-07 at 09 01 21](https://github.com/user-attachments/assets/ec2b17bb-c331-440c-9ae5-ab770c3062ab)

So probably `list_inference_profiles` is the correct way.

**Comment by longseespace** [comment]
OK I was able to use a cross-region model now. Will release soon!

![CleanShot 2024-12-07 at 15 58 57@2x](https://github.com/user-attachments/assets/322882d8-7a63-4052-8e0f-49643ef3b25a)

**Comment by longseespace** [comment]
I released v1.29.1. Give it a try.

**Comment by pstanisl** [comment]
It is working. Thank you.
